### ISSUE TITLE
Concurrent HDF5 reading and writing without SWMR

### ISSUE THREAD

**Issue by woutdenolf** [silx-kit/silx#3354]
For several reasons HDF5's SWMR is not usable:
1. No new groups/datasets can be created after the writer has enabled SWMR mode.
2. SWMR only works when the writer and the readers do things in a specific, synchronized order.
3. Non-SWMR readers (and that's basically most tools) will prevent the writer from enabling SWMR mode (or open the file to begin with). 
4. The SWMR feature does not work on network file systems (NFS, SMB/Windows file shares, and so on).

The alternative to SWMR is having a writer in normal mode (default h5py usage) and a reader that does not lock the file and captures exceptions caused by reading something that is being modified by the writer.

Proposal:`silx.io.h5py` to replace `h5py`

**Comment by woutdenolf** [comment]
I can derive a class from `h5py.File` to take care of the file locking, but a full h5py replacement doesn't seem possible. The h5py classes are clearly not designed to be derived from. Also closing and opening an HDF5 file for retry-on-exception invalidates all File, Group and Dataset instances. This makes a generic catch-retry very difficult.

All I can think of is to make decorator that does this:

```python
from silx.io import h5py

@h5py.retry(timeout=None)
def my_user_function(h5file):
     return list(h5file["/"])

scans = my_user_function("test.h5")
```

The `retry` decorator replaces the string with a `File` object.

**Comment by t20100** [comment]
This is what I asked this morning, whether we want/need to wrap only `h5py.File` or all `h5py` objects which is more complicated.
I understood it was only the `File` object so it is simpler but you don't catch everything.

Else you have to provide wrappers that provide wrappers.
In this case, there is https://github.com/silx-kit/silx/blob/master/silx/io/commonh5.py and also this project: https://github.com/clbarnes/h5py_like

**Comment by payno** [comment]
From my point of view having a `h5py.File` is already a first good start. On maintenance point of view dealing with the entire 'h5py' module would come with a higher cost.

**Comment by woutdenolf** [comment]
Yes I gave up on deriving from the `h5py` classes. I also thought to make a `commonh5` for `h5py` but it seems this would parse the entire file, or am I mistaken?

**Comment by t20100** [comment]
You should have some lazy-loading support (https://github.com/silx-kit/silx/blob/master/silx/io/commonh5.py#L1003).
But wrapping only `h5py.File` sounds better for now.

**Comment by woutdenolf** [comment]
One thing that I thought about `os.environ["HDF5_USE_FILE_LOCKING"]=...`: I cannot restore the old value after opening the file because there might be external links (typically when you open master files). So I could do this:
- Only restore the old value when all files are closed.
- When one file is open, you can only open another file with the same `HDF5_USE_FILE_LOCKING` value. Otherwise you'll get an exception.

This is not bullet proof because you could close a file like this `some_hdf5_group.file.close()` which won't free up choosing `HDF5_USE_FILE_LOCKING` for opening the next file. Acceptable?

**Comment by vasole** [comment]
What is wrong about just making sure is that when opening a file in any other mode than read only the variable is set to TRUE?

**Comment by vincefn** [comment]
What about the option of making the files read-only to all except the nexus writer ? 
It would require having a special user but it would be a foolproof method to avoid any issue with the writer, and we could all forget about this horrible LOCK variable for good.

**Comment by vasole** [comment]
That should not be a problem for the acquisition and I am in favor of it.

However, we should aim at providing some safety for other processes.

**Comment by woutdenolf** [comment]
> What is wrong about just making sure is that when opening a file in any other mode than read only the variable is set to TRUE (and to FALSE in the other cases)?

Because when following external links, new HDF5 files are being opened and they need to be opened with the same HDF5_USE_FILE_LOCKING value.

**Comment by vasole** [comment]
We are talking about making thing simple for people. I really doubt their intention is to write to a file via external links, therefore 99% of the cases are handled by that very simple approach and the goal of protecting and not blocking files is achieved.

**Comment by vincefn** [comment]
I agree - are there many cases where the main nexus output uses virtual datasets ? And are the links added before the linked file is actually closed ?
We're never going to cover every case - 99% will have to be enough. Unless the readonly permissions can make that 100%.

**Comment by vasole** [comment]
At this point there are mainly two uses of virtual data sets:

- To present the LIMA acquisitions from a single entry point instead of accessing the individual files (containing typically 100 frames)
- To create NXdata plots reshaping (n_points, data) into (rows, columns, data)

I do not think those are standard user tasks.

**Comment by woutdenolf** [comment]
> are there many cases where the main nexus output uses virtual datasets

When analyzing Bliss data, these actions will cause h5py to open additional files:
- Access lima images through any of the Nexus writer files (virtual datasets).
- Access any data through the sample or proposal master HDF5 files (external links).

VDS for NXdata is supported by the Nexus writer but not yet used in Bliss.

**Comment by maurov** [comment]
> 
> 
> This is what I asked this morning, whether we want/need to wrap only `h5py.File` or all `h5py` objects which is more complicated.
> I understood it was only the `File` object so it is simpler but you don't catch everything.
> 
> Else you have to provide wrappers that provide wrappers.
> In this case, there is https://github.com/silx-kit/silx/blob/master/silx/io/commonh5.py and also this project: https://github.com/clbarnes/h5py_like
> 
> You should have some lazy-loading support (https://github.com/silx-kit/silx/blob/master/silx/io/commonh5.py#L1003).
> But wrapping only `h5py.File` sounds better for now.

As user of `silx.io.open` I vote for the wrapping solution (`silx.io.h5h5`?). It would be very beneficial to get a `commonh5.File/Group` object when reading a HDF5 file. This would permit  `silx.io.open` to be consistent when reading data from other sources (Spec, Numpy, EDF, whatever). It will also permit writing to HDF5 directly from `silx`, without using directly `h5py`. Here a minimal example:

```python
import numpy as np
import h5py
import tempfile
from silx.io import open as silx_open
from silx.io.convert import write_to_h5
from silx.io.convert import _is_commonh5_group, is_group, is_dataset

#: generate dummy data in HDF5 file
ft_h5 = tempfile.mktemp(suffix='.h5')
hf = h5py.File(ft_h5, mode='a')
dat = np.random.normal(size=100)
hf.create_group('raw', track_order=True)
hf.create_dataset('/raw/1.1', data=dat)
hf.create_dataset('/raw/2.1', data=dat**2)
hf.close()

#: open with SILX
sf_h5 = silx_open(ft_h5)
sg = sf_h5['/raw/1.1']
print("is_group(sg):", is_group(sg))
print("is_dataset(sg):", is_group(sg))
print("_is_commonh5 group(sg):", _is_commonh5_group(sg))

#: writing to a second HDF5 file (fails!)
ft2 = tempfile.mktemp(suffix='.h5')
hf2 = h5py.File(ft2, mode='a')
write_to_h5(sg, hf2, h5path='/org/', overwrite_data=True)
hf2.close()
```

**Comment by t20100** [comment]
Regarding the file locking, there is a `H5Pset_file_locking` function in libhdf5 (https://github.com/HDFGroup/hdf5/blob/c256612e0950f3389249aa3a151d66ce01704d89/src/H5Pfapl.c#L4606)

According to the release notes, it is available in:
- v1.10.7 released on 2020-09-11 (https://github.com/HDFGroup/hdf5/blob/1.10/master/release_docs/RELEASE.txt#L113)
- 1.13 branch (https://github.com/HDFGroup/hdf5/blob/develop/release_docs/RELEASE.txt#L203)
- It is **NOT** (yet?) available in a 1.12 version

This is not (yet) available through `h5py`.

**Comment by woutdenolf** [comment]
We have two problems:

1. HDF5_USE_FILE_LOCKING=FALSE reader does not work when writer does `libver="latest"` (which means HDF5 1.10 for h5py 2.x or HDF5 1.12 for h5py 3.x)
2. `mode="w"` corrupts the file when already open is w or a mode, despite the HDF5 lock.

```python
import os
import h5py
from time import sleep
import multiprocessing


print(h5py.version.version)
print(h5py.version.hdf5_version)
assert os.environ.get("HDF5_USE_FILE_LOCKING") is None


def writer1(filename, **kw):
    os.environ["HDF5_USE_FILE_LOCKING"] = "TRUE"
    with h5py.File(filename, **kw) as f:
        f["check"] = True
        f.flush()
        print("writer1 holds file open")
        sleep(1000)


def writer2(filename, **kw):
    os.environ["HDF5_USE_FILE_LOCKING"] = "TRUE"
    try:
        with h5py.File(filename, **kw) as f:
            pass
    except Exception as e:
        print("writer2 failed as expected")
        pass  # locked by writer1


def reader(filename, **kw):
    os.environ["HDF5_USE_FILE_LOCKING"] = "FALSE"
    with h5py.File(filename, mode="r", **kw) as f:
        assert f["check"][()]
    print("HDF5 valid")


filename = "test.h5"
try:
    os.unlink(filename)
except FileNotFoundError:
    pass


if True:
    libver = None
else:
    # This does not allow HDF5_USE_FILE_LOCKING=FALSE reader
    libver = "latest"

p = multiprocessing.Process(
    target=writer1, args=(filename,), kwargs={"mode": "a", "libver": libver}
)
p.start()
sleep(1)
try:
    reader(filename)
    if True:
        writer2_mode = "a"
    else:
        # This corrupts the file, despite the lock!!!
        writer2_mode = "w"
    writer2(filename, mode=writer2_mode, libver=libver)
    reader(filename, libver=libver)
finally:
    p.kill()
```

**Comment by vasole** [comment]
Does that mean we have to avoid the use of the "w" mode in favor of the "a" mode?

**Comment by woutdenolf** [comment]
Yes, we should avoid "w" mode it seems.

To clarify, this sequence of event corrupts the file, despite the file locking:
1. writer1: opens the file in "a" or "w" mode (HDF5_USE_FILE_LOCKING=TRUE)
2. writer2: tries to open the file in "w" mode (HDF5_USE_FILE_LOCKING=TRUE) and fails

This sequence of event does not corrupt the file:
1. writer1: opens the file in "a" or "w" mode (HDF5_USE_FILE_LOCKING=TRUE)
2. writer2: tries to open the file in "a" mode (HDF5_USE_FILE_LOCKING=TRUE) and fails

**Comment by woutdenolf** [comment]
I will report the "w" mode issue. However the biggest problem is `liber="latest".

**Comment by woutdenolf** [comment]
Reported: https://github.com/h5py/h5py/issues/1804

Nevertheless I will proceed with this MR.

**Comment by vincefn** [comment]
> 1. A concurrent reader with `HDF5_USE_FILE_LOCKING=FALSE` does not work when the writer creates the file with `libver="latest"` (which means HDF5 1.10 for h5py 2.x or HDF5 1.12 for h5py 3.x)

I don't understand - what changed ? Reading without the lock is precisely what was working with libver='latest' ?

> 2. `mode="w"` corrupts the file when already open is w or a mode, despite the HDF5 lock.

Is there any case where we need two writers ? If this is just about rogue opening in 'a' mode, the permissions would solve this.

**Comment by woutdenolf** [comment]
> > 1. A concurrent reader with `HDF5_USE_FILE_LOCKING=FALSE` does not work when the writer creates the file with `libver="latest"` (which means HDF5 1.10 for h5py 2.x or HDF5 1.12 for h5py 3.x)
> 
> I don't understand - what changed ? Reading without the lock is precisely what was working with libver='latest' ?

It was working in the test you are referring to because the retry loop didn't stop as long as the writer was holding the file open. The writer releases the file between scans. With `libver=None` does get hold of the file and the top-level groups but processing those top-level groups fails, although these scans are finished (something to do with H5_ITERATE, lets not discuss that here).

> > 1. `mode="w"` corrupts the file when already open is w or a mode, despite the HDF5 lock.
> 
> Is there any case where we need two writers ? If this is just about rogue opening in 'a' mode, the permissions would solve this.

No we never need two writers. This point is related to safety against file corruption. If someone is opening the file in w mode (for whatever reason) the file is corrupt. In other word the HDF5 file lock apparently does not protect against this kind of malicious attack on the file.

**Comment by vincefn** [comment]
OK - so in both cases, having a readonly permission would solve the issue - no rogue writer, and the reader cannot lock in any way the file.

But how does the reader get hold of the file ? As far as I could see reading without a lock does _not_ modify the file ? At least I have no problem opening a file for reading without lock first, and then opening to append, but that's just a manual test.

And in the test the writer was closing and re-opening the file every time, it writes:
``` python
def writer(filename, n=20):
   for j in range(n):
        ...
        with h5py_utils.File(filename, mode="a", libver="v110") as f:
           ...
```

**Comment by woutdenolf** [comment]
> OK - so in both cases, having a readonly permission would solve the issue - no rogue writer, and the reader cannot lock in any way the file.

No you misunderstand. The first issue is: writer opens and locks file, reader can no longer open the file with HDF5_USE_FILE_LOCKING=FALSE. It can only do that with `libver=None`.

**Comment by woutdenolf** [comment]
> But how does the reader get hold of the file ? 

When the writer closes the file, which happens after every scan.

**Comment by woutdenolf** [comment]
So to be clear, we don't gain anything related to concurrent reading and writing with `libver="v110"` or `libver="latest"` or whatever

**Comment by vincefn** [comment]
Sorry but I still don't understand - in the safe_open_hdf5.py test the readers are constantly reading the files, while the writer continues to loop {open the file in append mode, add something, close the file} - and the writer never had any issue with that concurrent access.

I don't understand what you are losing.

**Comment by woutdenolf** [comment]
writer:
1. open/lock(HDF5_USE_FILE_LOCKING=TRUE, mode="a", libver=None)
2. write data
3. close
reader:
1. open(HDF5_USE_FILE_LOCKING=FALSE, mode="a", libver=None)
2. get all finished scan groups (all or all but the last)
3. process the scan data
4. close

So the writer will never be disturbed. The reader will have to retry step 1 and 2 until in succeeds (I will say that it "gets hold" of the file). So far so good, but there is an unexpected caveat, when the reader succeeds in step 1 and 2, step 3 can fail.

So then we noticed that with `libver="v110"` step 3 never fails

writer:
1. open/lock(HDF5_USE_FILE_LOCKING=TRUE, mode="a", libver="v110")
2. write data
3. close
reader:
1. open(HDF5_USE_FILE_LOCKING=FALSE, mode="a", libver="v110")
2. get all finished scan groups (all or all but the last)
3. process the scan data
4. close

However why does it never fail? Because reader step 1 fails as long as the writer is accessing the file (i.e. during a scan). This is not achieved by the lock (because locking is disabled on the reader side). So the reader which is frantically looping to "get hold" of the file succeeds only when the writer is not accessing the file.

In other words, by setting `libver="v110"` we accidentally found a way to avoid concurrent access without having the reader lock the file.

That's great but this means the reader has only a short time in between scans (when the writer releases the file) to "get hold" of the file. So we kind of get concurrent data processing but not really.

So we are loosing:
* data processing on the current acquisition file (unless you can "get hold" of the file between scans)
* we loose support for Matlab, IDL, etc. (all things which ship with HDF5 1.8.x)

**Comment by vincefn** [comment]
Thanks, that's more clear.

> That's great but this means the reader has only a short time in between scans (when the writer releases the file) to "get hold" of the file. So we kind of get concurrent data processing but not really.

Yes. the question is how does that fit with actual data collection and whether we can live with that - it seems to fit high-throughput/large number of datasets acquisitions but it won't fit all...

> So we are loosing:
> * data processing on the current acquisition file (unless you can "get hold" of the file between scans)

As far as I can see we did not have the ability to do data processing, and we can do that now, even if it has the limitations you mention. So not a loss, just not a total gain (and with some uncertainties because the success is indeed accidental).

> * we loose support for Matlab, IDL, etc. (all things which ship with HDF5 1.8.x)

In the end it may be up to beamlines to decide whether they want to use matlab/idl for analysis during the experiment, and use v110 or not. After that for users there's `h5repack` and `h5format_convert` (we don't have the latter on the cluster ?)

No perfect solution, but as long as the writer regularly closes the file this should be a gain for a large percentage of beamlines.

**Comment by vasole** [comment]
Then the 1.10 solution is the user refreshes the HDF5 file between scans. That is easy and it is equivalent to accessing all the scans but the last one that can in any case be accessed via redis. 

However, I do not like the incompatibility with MatLab. Repack will duplicate files and when using Virtual Data Sets those files will be huge. If we could get a python access to HDF5 using a statically linked HDF5 library all MatLab problems would be solved. It is straightforward to call Python from MatLab but there is a mess with dynamically linked libraries because MatLab and h5py use different versions of the same library.

I am almost sure I can achieve the same thing while remaining in 1.8 version if the user does not try to refresh the HDF5 while being updated. PyMca has been working like that since long time and without automatic retry on failure.

**Comment by vincefn** [comment]
Apparently `h5format_convert` (I have not tested, it seems to be missing in ubuntu - https://bugs.launchpad.net/ubuntu/+source/hdf5/+bug/1860874) only repacks the metadata in the file, so it's not so bad.

Ultimately we just have to give a few options to the beamlines. Those using a 'slow' acquisition which produces independent data files won't really care, they can wait for each dataset to be closed. But for high throughput multi-datasets like John's example it's good to have the option to perform some analysis while the file is still written.

**Comment by vasole** [comment]
For high throughput I still maintain that LIMA files are already split and 0d data should be easily accessible from redis.

As a last resort we still have the possibility to split the NexusWriter file each n scans in a subdirectory.

If one really wants to work online, it is not from files that the information has to be obtained but from memory.

**Comment by vincefn** [comment]
There are different degrees of 'online' work - the instantaneous feedback should work from memory, sure - but that requires a suitable API for access, and will probably not be easy enough for users, so it'll require much more human resources for development and is only a partial solution.

The control workstations are not built for real analysis work - online analysis has to be possible from a different computer, notably the jupyter-slurm cluster. That may happen with minutes delay but it still online.

**Comment by vasole** [comment]
> The control workstations are not built for real analysis work - online analysis has to be possible from a different computer, notably the jupyter-slurm cluster. That may happen with minutes delay but it still online.

An so what? If it is a question of minutes you should be able to access REDIS from the slurm cluster and the LIMA data have no problem at all. Just install the bliss packages in a slurm conda environment and you have it. 0d-data were never generating GBytes in a question of minutes (we never got such things with SPEC files).

**Comment by stufisher** [comment]
Bliss only holds 2000 blocks (points) of data in redis for a particular scan node, at this point you cant get the data back from redis and it must be accessed from file. We hit this limit regularly on id21 and i have to switch from generating maps in-memory to generating from h5 file (spectra data) during a scan.

I am surprised about:
> That's great but this means the reader has only a short time in between scans (when the writer releases the file) to "get hold" of the file. So we kind of get concurrent data processing but not really. If it manages to get hold of the file, it can process data while the writer starts writing the next scan. So it is concurrent in that way, but it has to get lucky to "get hold" of the file first.

We have been using this methodology for about a year and still managed to read the h5 from time to time to regenerate xrf maps in the ui from within a large, say 20000-50000 point scan. I guess we have just been lucky that the writer is not holding the file at that time, even though its the same scan?

**Comment by woutdenolf** [comment]
> We have been using this methodology for about a year and still managed to read the h5 from time to time

Sure, I'm not saying it never works. You can manage to read the data from time to time, even from the ongoing scan.

I will add utility methods to silx to facilitate exception catch/retry (which means re-open the file) in case HDF5 IO does fail. I'm trying to find when I can expect exceptions. For example, what I was hoping for was: when I get hold of the NXentry HDF5 group of a finish scan (possibly after several catch/retry), I can safely process that scan. But this doesn't seem to be the case and I want to know whether that is inherent to the HDF5 structure or whether it is caused by unnecessary HDF5 header/heap/root-b-tree access when doing certain operations.

Apart from exceptions we also have segfaults, which is another level of problematic.
